From fdf9a313361032f5524bcbb6afd964b95c900415 Mon Sep 17 00:00:00 2001
From: Ondrej Jirman <megi@xff.cz>
Date: Fri, 8 Aug 2025 17:54:21 +0200
Subject: Revert "ASoC: soc-core: merge snd_soc_unregister_component() and
 snd_soc_unregister_component_by_driver()"

This reverts commit 144d6dfc7482455eabf8e8caa974a6e8d9572705.
---
 include/sound/soc.h  |  2 +-
 sound/soc/soc-core.c | 30 +++++++++++++++++++++++++-----
 2 files changed, 26 insertions(+), 6 deletions(-)

diff --git a/include/sound/soc.h b/include/sound/soc.h
index 1fffef311c41..8d113ee8c2bc 100644
--- a/include/sound/soc.h
+++ b/include/sound/soc.h
@@ -444,7 +444,7 @@ int snd_soc_register_component(struct device *dev,
 int devm_snd_soc_register_component(struct device *dev,
 			 const struct snd_soc_component_driver *component_driver,
 			 struct snd_soc_dai_driver *dai_drv, int num_dai);
-#define snd_soc_unregister_component(dev) snd_soc_unregister_component_by_driver(dev, NULL)
+void snd_soc_unregister_component(struct device *dev);
 void snd_soc_unregister_component_by_driver(struct device *dev,
 			 const struct snd_soc_component_driver *component_driver);
 struct snd_soc_component *snd_soc_lookup_component_nolocked(struct device *dev,
diff --git a/sound/soc/soc-core.c b/sound/soc/soc-core.c
index d31ee6e9abef..222edc0b433a 100644
--- a/sound/soc/soc-core.c
+++ b/sound/soc/soc-core.c
@@ -2931,14 +2931,34 @@ EXPORT_SYMBOL_GPL(snd_soc_register_component);
 void snd_soc_unregister_component_by_driver(struct device *dev,
 					    const struct snd_soc_component_driver *component_driver)
 {
-	const char *driver_name = NULL;
+	struct snd_soc_component *component;
 
-	if (component_driver)
-		driver_name = component_driver->name;
+	if (!component_driver)
+		return;
 
+	mutex_lock(&client_mutex);
+	component = snd_soc_lookup_component_nolocked(dev, component_driver->name);
+	if (!component)
+		goto out;
+
+	snd_soc_del_component_unlocked(component);
+
+out:
+	mutex_unlock(&client_mutex);
+}
+EXPORT_SYMBOL_GPL(snd_soc_unregister_component_by_driver);
+
+/**
+ * snd_soc_unregister_component - Unregister all related component
+ * from the ASoC core
+ *
+ * @dev: The device to unregister
+ */
+void snd_soc_unregister_component(struct device *dev)
+{
 	mutex_lock(&client_mutex);
 	while (1) {
-		struct snd_soc_component *component = snd_soc_lookup_component_nolocked(dev, driver_name);
+		struct snd_soc_component *component = snd_soc_lookup_component_nolocked(dev, NULL);
 
 		if (!component)
 			break;
@@ -2947,7 +2967,7 @@ void snd_soc_unregister_component_by_driver(struct device *dev,
 	}
 	mutex_unlock(&client_mutex);
 }
-EXPORT_SYMBOL_GPL(snd_soc_unregister_component_by_driver);
+EXPORT_SYMBOL_GPL(snd_soc_unregister_component);
 
 /* Retrieve a card's name from device tree */
 int snd_soc_of_parse_card_name(struct snd_soc_card *card,
-- 
2.51.0

